src/builder/guard.rs

```diff
@@ -109,3 +143,41 @@ impl<'a> Guard<'a> {
         return Ok(());
     }
 }
+
+fn remove_all_except(root: &Path, keep_rel_paths: &HashMap<&Path, bool>)
+    -> Result<(), String>
+{
+    let entries = try_msg!(root.read_dir(),
+        "Can't read dir {dir:?}: {err}", dir=root);
+    for entry in entries {
+        let ref path = try_msg!(entry,
+                "Can't iterate over dir entries {dir:?}: {err}",
+                dir=root)
+            .path();
+        let ref rel_path = path.strip_prefix("/vagga/root").unwrap();
+        match keep_rel_paths.get(rel_path) {
+            Some(&is_final) if is_final => {
+                continue;
+            },
+            Some(_) => {
+                try!(remove_all_except(path, keep_rel_paths));
+            },
+            None => {
+                if path.is_dir() {
+                    try_msg!(clean_dir(path, true),
+                        "Error cleaning dir {path:?}: {err}",
+                        path=path);
+                } else {
+                    try_msg!(remove_file(path),
+                        "Error removing file {path:?}: {err}",
+                        path=path);
+                }
+            },
+        }
+    }
+    Ok(())
+}
+
+fn warn_duplicate_data_dir(path: &Path) {
+    warn!("{:?} is already marked as data directory", Path::new("/").join(path));
```

Comment:
The warning is a little bit misleading:
```
[/var/log, /var]
->  "/var" is already marked as data directory
```
Well, it's not.

You can use a boolean value from `keep_rel_paths` to refine warning. Or you can describe both cases, like ".. is either a prefix of other directory or already contained in another prefix"